### PR TITLE
Initial integration

### PR DESCRIPTION
**Pull request by Edke**
Initial integration of tap:
- python package skeleton
- cli and config loading
- base objects Tap and Stream, separation of base functionality from Pipedrive specific functionality, in `tap_pipedrive.singer`
- PipedriveTap and PipedriveStream objects, Pipedrive related integration within these
- Streams Currencies, Deals and Notes
- Notes are not in specs, only added as was easily populated with data for pagination integration
- Response validation
- Pagination

If design accepted, we can refactor and integrate base classes within singer.io lib.

### CODE REVIEW

**Comment by Edke** [comment]
tap config example:
```
{
  "api_token": "XXXXXXXXXXXXXXXXXXXXXX",
}
```

**Comment by cmerrick** [comment]
This looks like a great start.

I see what you did with the generic `Tap` and `Stream` base classes, and it makes sense, but at this point we're not ready to add those concepts to `singer-python`, so you should just design for this specific tap.  Our rationale is that putting those objects into `singer-python` will force a tap writer to use OOP throughout, and we don't really think that's necessary in all taps.

A few notes for further development:
 - Please instrument your code with the singer metrics library following these examples: https://github.com/singer-io/tap-xero/blob/562d32d9fcc3b52691fe6cb7ed110596c9dcb25d/tap_xero/__init__.py#L113 and https://github.com/singer-io/tap-xero/blob/562d32d9fcc3b52691fe6cb7ed110596c9dcb25d/tap_xero/pull.py#L16
 - This code won't be licensed under the MIT license, please remove that from setup.py

Looking forward to seeing the rest!

File: tap_pipedrive/tap.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        RecentNotesStream(),
+        RecentUsersStream(),
+        RecentActivitiesStream(),
+        RecentDealsStream(),
+        RecentFilesStream(),
+        RecentOrganizationsStream(),
+        RecentPersonsStream(),
+        RecentProductsStream(),
+        RecentDeleteLogsStream()
+        # RecentStagesStream(),
```
[the remark this answers is not in the document]
Comment (Edke) [reply tap_pipedrive/tap.py]: @cmerrick these two streams were wrongly integrated, `Stage` is a recent version and direct `/stage/` endpoint should be used and vice versa `Notes` should be used `/recent/` version.

Should I remove them both Stream() classes and schemas ?

File: tap_pipedrive/config.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+BASE_URL = "https://companydomain.pipedrive.com/v1"
```
[the remark this answers is not in the document]
Comment (Edke) [reply tap_pipedrive/config.py]: @cmerrick shouldn't SSL be used ? According to API: `All requests to the API must be made over SSL (https not http).`

[the remark this answers is not in the document]
Comment (cmerrick) [reply tap_pipedrive/config.py]: yes, absolutely, typo by me: `https://api.pipedrive.com/v1`

**Comment by Edke** [comment]
hello @cmerrick, was trying to go though all your comments, will re-write them as separate tasks so we can easily track ones that I fixed and what is remaining

**Comment by Edke** [comment]
- [ ] use the exact names from the project plan document. Specifically, activityTypes should be activity_types, deleteLog should be delete_log
- [ ] for data from all of the recents endpoints, only write the contents of the data object (or the first object in the data array for items=user) - so if you're currently writing a python dictionary "o", you should instead write "o['data']", and for the items=user case, you should write "o['data'][0]".
- [ ] This part of the project plan isn't implemented: In addition to a bookmark for each endpoint, the tap should also maintain state indicating which stream it is currently syncing, which is cleared when the tap completes successfully. When the tap is invoked and this value is set, it should resume from this stream. This ensures that if the tap is terminated before it completes syncing all streams, when it starts again it can resume from the last stream it was syncing.
- [ ] In stream.py - the call to singer.metrics.http_request_timer should be wrapping the HTTP call, which isn't the case currently, so it's returning inaccurate metrics
- [ ] the goals and users streams are not updating the state dictionary. Also, when tap is run with a state file but no entries for these two streams, no data is emitted for either of the two streams.
- [ ] custom fields (e.g. 6c6ad03c59918d636b9b0a2a0b263a89213bf4a4 for deals) shouldn't be statically saved into the schema json files, they should be added dynamically because they will be different for every pipedrive account. The project plan notes which endpoints support this dynamic typing - activities, deals, notes, organizations, persons, products. These streams have supplementary API endpoints for discovering which fields are available and their types, and I would expect the schemas to be generated dynamically from these endpoints rather than from static files
- [ ] when I run the tap in the test account without a state and with a start_time of 2017-09-01T00:00:00Z, I don't seem to be getting all of the notes, e.g. notes id 301+. When I run it a second time and pass it the state from the first run, I get them. Possibly a pagination issue?
- [ ] One other note that's not related to the function of the tap - the quality of this code would be higher if you merged the base classes in the singer/ folder with their child classes in the root folder. Those base classes are not generically applicable to all singer taps for a few reasons (for example, not all taps will write state at the end of each stream), and they made this code much more difficult for me to read because I had to follow logic back and forth between each base class and its child class.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+from tap_pipedrive import __version__
+
+
+setup(
+    name="tap-pipedrive",
+    version=__version__,
+    license="AGPL",
+    description="Singer.io tap for extracting data from the Pipedrive API",
+    author="Stitch",
+    author_email="eduard.kracmar@hotovo.org",
```
[the remark this answers is not in the document]
Comment (Edke) [reply setup.py]: @cmerrick fixed in 52d46a9

File: tap_pipedrive/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Edke) [reply tap_pipedrive/config.py]: Fixed in 3e2742d

File: tap_pipedrive/streams/recents/users.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from tap_pipedrive.streams.recents import RecentsStream
+
+
+class RecentUsersStream(RecentsStream):
+    items = 'user'
+    schema = 'users'
+    key_properties = ['id', ]
+    # disabled bookmarking till settled it's data / array
```
[the remark this answers is not in the document]
Comment (Edke) [reply tap_pipedrive/streams/recents/users.py]: Fixed in 3dd28ae

File: tap_pipedrive/tap.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        return requests.get(url, headers=headers, params=params)
+
+    def validate_response(self, response):
+        if isinstance(response, requests.Response) and response.status_code == 200:
+            try:
+                payload = response.json()
+                if payload['success'] and 'data' in payload:
+                    return True
+
+            # TODO narrow down
```
[the remark this answers is not in the document]
Comment (Edke) [reply tap_pipedrive/tap.py]: fixed in 23febac

File: tap_pipedrive/tap.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmerrick) [reply tap_pipedrive/tap.py]: yes, please remove

**Comment by Edke** [comment]
>  the goals and users streams are not updating the state dictionary. Also, when tap is run with 
> a state file but no entries for these two streams, no data is emitted for either of the two streams.

**goals** endpoint has disabled bookmarking as no `update_time`, we don't have a timestamp to filter records, therefore on each run all records have to be displayed

**users** endpoint currently waiting for response from Pipedrive devs as unpredictable behavior, currently it seems that `since_timestamp` is compared to `created` field and not `modified`.

**Comment by Edke** [comment]
All requested changes should be handled, more info in e-mail.

**Comment by Edke** [comment]
> In the goals schema, expected_sum is an integer, but delivered_sum is a number - I suspect these 
> should have the same type. Do you know which is correct?


Only 4 goals in my test account in this endpoint. 

`expected_sum`: 0 or 1
`delivered_sum`: 0

So from these values both are integers. Should be tested against production data. But I guess these mean currency so changing both to `number` should work.

**Pull request merged**